File: algorithm/백준/1405.cpp
#include <stdio.h>
int N;
double h[4];
bool vl[30][30] = { 0, };
double res = 0.0;
char D[2][4] = { 1, -1, 0, 0, 0, 0, 1, -1 };

void dfs(double cur_h = 1.0, int x = 14, int y = 14, int dep = 0) {
	// 재방문 없이 횟수 다 한 경우..
	if (dep == N) {
		res += cur_h;
		return;
	}

	int tx, ty;
	for (int d = 0; d < 4; d++) {
		tx = x + D[0][d];
		ty = y + D[1][d];
		if (!vl[ty][tx]) {
			vl[ty][tx] = 1;

			dfs(cur_h * h[d], tx, ty, dep + 1);
			vl[ty][tx] = 0;
		}
	}

}
int main(void) {
	int tmp;
	scanf("%d", &N);
	for (int t = 0; t < 4; t++) {
		scanf("%d", &tmp);
		h[t] = tmp / 100.0;
	}
	vl[14][14] = 1;
	dfs();
	printf("%.10lf\n", res);
	return 0;

}

File: algorithm/백준/1405.py
vl = [[0 for _ in range(30)] for _ in range(30)]
res = 0.0
D = [[1, 0, 0], [-1, 0, 1], [0, 1, 2], [0, -1, 3]]
cur_h = 1
dep = 0
def dfs(x=14, y=14):
    global cur_h, dep
    if dep == N:
        global res
        res += cur_h
        return
    for dx, dy, d in D:
        tx = x + dx
        ty = y + dy
        if not vl[ty][tx] and h[d] != 0:
            vl[ty][tx] = 1
            cur_h *= h[d]
            dep += 1
            dfs(tx, ty)
            cur_h /= h[d]
            vl[ty][tx] = 0
            dep -= 1

input_val = list(map(int, input().split()))
N = input_val[0]
h = [a/100 for a in input_val[1:]]

vl[14][14] = 1
dfs()
print(res)

